text/0580-destroyables.md

```diff
@@ -0,0 +1,243 @@
+destructors to accomplish this goal:
+
+```js
+let parent1 = {}, parent2 = {}, child = {};
+
+registerDestructor(parent1, () => destroy(child));
+registerDestructor(parent2, () => destroy(child));
+```
+
+The exact timing semantics here will be a bit different, but for most use cases
+this should be fine. If we find that it would be useful to have multiple
+inheritance baked in in the future, it can be added in a followup RFC.
+
+#### `registerDestructor`
+
+Receives a destroyable and a destructor function, and associates the function
+with it. When the destroyable is destroyed, or when its parent is destroyed, the
+destructor function will be called. Multiple destructors can be associated with
+a given destroyable, and they can be associated over time, allowing libraries
+like `ember-lifeline` to dynamically add destructors as needed.
+
+- Registering a destructor on a destroyed object should throw an error.
+- Attempting to register the same destructor multiple times should throw an
+  error.
+- Attempting to register a destructor on a non-destroyable should throw an
+  error.
+
+#### `unregisterDestructor`
+
+Receives a destroyable and a destructor function, and de-associates the
+destructor from the destroyable.
+
+- Calling `unregisterDestructor` on a destroyed object should throw an error.
+- Calling `unregisterDestructor` with a destructor that is not associated with
+  the object should throw an error.
+- Calling `unregisterDestructor` on a non-destroyable should throw an error.
+
+#### `destroy`
+
+`destroy` manually initiates the destruction of a destroyable. This allows users
+to manually manage the lifecycles of destroyables that are shorter lived than
+their parents (for instance, destroyables that are defined an a service, which
+lives as long as an application does), cleaning them up as needed.
+
+Calling `destroy` multiple times on the same object is safe. It will not throw
+an error, and will not take any further action.
+
+- Calling `destroy` on a non-destroyable should throw an error.
+
+#### `isDestroying`
+
+This would return true as soon as the destroyable begins destruction, before any
+of its destructors are called. Returns false otherwise.
+
+- Calling `isDestroying` on a non-destroyable should throw an error.
+
+#### `isDestroyed`
+
+This would return true as soon after the destroyable has been fully destroyed,
+including all of its child destroyables. Returns false otherwise.
+
+- Calling `isDestroyed` on a non-destroyable should throw an error.
+
+#### `assertDestroyablesDestroyed`
+
+This function asserts that all active destroyables have been destroyed at the
+time it is called. It is meant to be a low level hook that testing frameworks
+like `ember-qunit` and `ember-mocha` can use to hook into and validate that all
+destroyables have in fact been destroyed.
+
+### Timing Semantics
+
+Destroyables first call their own destructors, then destroy their associated
+children.
```

Comment:
I'd like a bit more flesh here, and I think we need some tweaks. Specifically, I think that the order of operation should be:

1. mark the destroyable such that `isDestroying(destroyable)` returns `true`
2. call the destructors
3. mark _all_ children associated such taht `isDestroying(child)` returns `true`
4. call the destructors for the child
5. recurse.....